### PR TITLE
meson: optionally build seccomp if it supports notify

### PR DESCRIPTION
**Pull request by haircommander**
Signed-off-by: Peter Hunt <pehunt@redhat.com>

### CODE REVIEW

**Comment by haircommander** [comment]
theoretically this fixes https://github.com/containers/conmon/issues/337

File: meson.build [hunk 1]
```diff
@@ -45,12 +45,19 @@ endif
...
 sd_journal = dependency('libsystemd-journal', required : false)
 if not sd_journal.found()
-	sd_journal = dependency('libsystemd', required : false)
+  sd_journal = dependency('libsystemd', required : false)
 endif
 if sd_journal.found()
 	add_project_arguments('-DUSE_JOURNALD=1', language : 'c')
 endif
 
+if run_command('./hack/seccomp-notify.sh', check: false).returncode() == 0
```
Comment (eli-schwartz) [inline meson.build:54]: This relies on a shell script to do compile checks, which doesn't respect $CC (there may not necessarily be a `cc` on $PATH, though there likely is) or $PKG_CONFIG (which may point to `pkgconf`).

Maybe it should be ported to something Meson understands natively?

File: meson.build [hunk 2]
```diff
@@ -45,12 +45,19 @@ endif
...
+  seccomp = dependency('libseccomp', required : false)
```
Comment (eli-schwartz) [inline meson.build:55]: You don't seem to ever add this dependency anywhere?

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

**Comment by kloczek** [comment]
IMO that is not correct solution.
if libseccomp API is used independently from libsystemd is used simple in meson detection should be two checks for those two libraries.
How it was found in https://github.com/containers/conmon/issues/337 containers uses directly libseccomp API even if it was not possible to detect libsystemd  so meson detection should not rely on indirect dependencies 

Using some additional script to check availability of the `seccomp_notif_sizes ()` is as well incorrect because all what dhould be checked is just minimum version of the pkgconfig module.

**Comment by kloczek** [comment]
Just checked and it is not possible to build libseccomp without that `seccomp_notif_sizes ()` routine so checking it is pointless.

**Comment by haircommander** [comment]
> 

Actually , the only reason we link with libseccomp is to use seccomp notify, which requires that structure

File: meson.build (hunk 2 above)
Comment (haircommander) [reply meson.build]: fun fact: I've never used meson before! thanks for the heads up :-)

**Comment by haircommander** [comment]
I've updated with some suggestions, thanks for the reviews y'all

> Using some additional script to check availability of the seccomp_notif_sizes () is as well incorrect because all what dhould be checked is just minimum version of the pkgconfig module.

the problem is we need to know if seccomp notify is supported, but I believe packaged libseccomps don't necessarily correctly reflect that with pkg-config (for instance if support was backported) hence the additional check by compiling

**Comment by shenghaoyang** [comment]
Maybe it would be cleaner to use the built-in test features of meson? See https://mesonbuild.com/Compiler-properties.html - there's a brute force "does snippet compile" function and a separate helper for sizeof: https://mesonbuild.com/Compiler-properties.html#expression-size. Should do the trick for structure presence detection.

Bonus: since it uses the same C compiler meson is setup to use, there shouldn't be any issues with the check being done with one compiler and the build performed by another.

File: meson.build (hunk 1 above)
Comment (eli-schwartz) [reply meson.build]: That's not something Meson understands natively. Also that doesn't help if someone configures Meson using a machine file (e.g. cross-compilation).

**Comment by haircommander** [comment]
thanks for the help everyone, I have attempted with the sizeof function

**Comment by haircommander** [comment]
I think that frankensystem was ubuntu: https://github.com/containers/conmon/issues/275

since 16.04 is out of service, and ubuntu 18.04 is on its way, maybe we won't need it much longer

For now, risk seems low to keep it

File: meson.build [hunk 3]
```diff
@@ -45,12 +45,17 @@ endif
...
 sd_journal = dependency('libsystemd-journal', required : false)
 if not sd_journal.found()
-	sd_journal = dependency('libsystemd', required : false)
+  sd_journal = dependency('libsystemd', required : false)
 endif
 if sd_journal.found()
 	add_project_arguments('-DUSE_JOURNALD=1', language : 'c')
 endif
 
+seccomp = dependency('libseccomp', required : false)
```
[the remark this answers is not in the document]
Comment (haircommander) [reply meson.build]: fixed!

**Comment by rhatdan** [comment]
@haircommander What is going on with this?

**Comment by kloczek** [comment]
IM that PR soes not make any sense because notify is supported since exact version of the libseccomp so checking dependencies usimg only `dependency()` covers 100% what is needed.
I have my own version fix of that issue:
```patch
--- a/meson.build
+++ b/meson.build
@@ -33,6 +33,10 @@
                       language : 'c')

 glib = dependency('glib-2.0')
+seccomp = dependency('libseccomp', version : '>= 2.5.2')
+if seccomp.found()
+  add_project_arguments('-DUSE_SECCOMP=1', language : 'c')
+endif

 cc = meson.get_compiler('c')
 null_dep = dependency('', required : false)
@@ -85,7 +86,7 @@
             'src/utils.h',
             'src/seccomp_notify.c',
             'src/seccomp_notify.h'],
-           dependencies : [glib, libdl, sd_journal],
+           dependencies : [glib, libdl, sd_journal, seccomp]
            install : true,
            install_dir : join_paths(get_option('libexecdir'), 'podman'),
 )
```
With that patch hack/seccomp-notify.sh can be deleted.

**Comment by haircommander** [comment]
adapted as suggested. @rhatdan PTAL and I'll cut v2.1.3

**Comment by kloczek** [comment]
With that this shell backed sritpt can be removed as well 😋 (as no longer needed)

**Comment by haircommander** [comment]
I'd prefer to keep it for now as it's still used in the makefile :)

**Pull request merged**